src/validators/__test__/sequenceValidator.spec.js

```diff
@@ -0,0 +1,106 @@
+const validAminoAcidSequence = 'ARNDCEQGHILKMFPSTWYVXBZJ';
+const invalidAminoAcidSequence = 'ARNDCEQGHILKMFPSTWYVXBZJO';
+const validNucleicAcidSequence = 'ACGTUACGTUACGTUACGTU';
+const invalidNucleicAcidSequence = 'ACGTUACGTUACGTUACGTU_';
+const validAmbiguousNucleicAcidSequence = 'ACGTWSMKRYBDHVNZ';
+const junkSequence = 'A1!B2"C3£D4$F5%G6^H7&';
+const withFASTA = `
+>FBgn0046814 type=gene; loc=2L:complement(9950437..9950455); ID=FBgn0046814; name=mir-87; dbxref=FlyBase:FBgn0046814,FlyBase:FBan0032981,FlyBase_Annotation_IDs:CR32981,GB:AE003626,MIR:MI0000382,flight:FBgn0046814,flymine:FBgn0046814,hdri:FBgn0046814,ihop:1478786; derived_computed_cyto=30F1-30F1%3B Limits computationally determined from genome sequence between @P{EP}CG5899<up>EP701</up>@ and @P{EP}CG4747<up>EP594</up>@%26@P{lacW}l(2)k13305<up>k13305</up>@; gbunit=AE014134; MD5=10a65ff8961e4823bad6c34e37813302; length=19; release=r5.21; species=Dmel;
+ACGTU
+>AT1G01140.1 (version 1)
+ACGTU
+>AT1G01140.1 (version 2)
+ACGTU
+>AT1G01140.2 (version 1)
+ACGTU
+>AT1G01140.2 (version 1)
+ACGTU
+`;
+
+beforeEach(() => {
+  results = null;
+  expectedResult = null;
+});
+
+test('validation should fail when sequence is missing', () => {
+  results = sequenceValidator(missingSequence);
+  expect(results)
+    .toEqual(errorResponses.missingSequence);
+});
+
+test('validation should fail when sequence is an empty string', () => {
+  results = sequenceValidator(emptySequence);
+  expect(results)
+    .toEqual(errorResponses.missingSequence);
+});
+
+test('should fail if a sequence is too short after clean-up', () => {
+  results = sequenceValidator(junkSequence);
+  expect(results)
+    .toEqual(errorResponses.invalidSequence);
+});
+
+test('should indicate a valid AA sequence', () => {
+  results = sequenceValidator(validAminoAcidSequence);
+  expectedResult = {
+    ...validResponse,
+    likelyType: 'aa',
+  };
+
+  expect(results)
+    .toEqual(expectedResult);
+});
+
+test('should indicate invalid for an AA sequence with invalid characters', () => {
+  results = sequenceValidator(invalidAminoAcidSequence);
+  expect(results)
+    .toEqual(errorResponses.invalidSequence);
+});
+
+test('should indicate a valid NA sequence', () => {
+  results = sequenceValidator(validNucleicAcidSequence);
+  expectedResult = {
+    ...validResponse,
+    likelyType: 'na',
+  };
+
+  expect(results)
+    .toEqual(expectedResult);
+});
+
+test('should indicate invalid for an NA sequence with invalid characters', () => {
+  results = sequenceValidator(invalidNucleicAcidSequence);
+  expect(results)
+    .toEqual(errorResponses.invalidSequence);
+});
+
+test('should indicate AA for an NA sequence with ambiguous codes', () => {
+  results = sequenceValidator(validAmbiguousNucleicAcidSequence);
+  expectedResult = {
+    ...validResponse,
+    likelyType: 'aa',
+  };
+
+  expect(results)
+    .toEqual(expectedResult);
+});
+
+test('should remove all FASTA description lines', () => {
```

Comment:
So it's removing all the comments and considering all of the sequences as one to calculate type likelihood, right?
I don't know if we might rather want to test sequence by sequence, independently, and return an array of likelihood results.

Reply:
Yes, that FASTA sample is just basically to try different headers and see if the RegEx works fine on different lines. The requirement doesn't say "handle multiple sequences" and the last time I used recursion, it was removed to make the code simpler. So, I am somewhat reluctant to support multiple sequences, although it doesn't need anything like recursion.

Do you think this function should be able to handle multiple sequences?

Reply:
I don't know to be honest, because the blast server doesn't handle multiple sequences out of the box anyway, maybe check with the others.
My main point would be that it's doing something quite unexpected to me. Assume I put a short first sequence that would be rejected on its own, but a valid long 2nd sequence. Then this would tell me this is valid (because it's merging both sequences).
In this case I would expect it to either:
 - Test only the first sequence and fail (because too short), and ignore the 2nd
 - Test both sequences and return an array of [<failed>, <ok>]
 - Fail because there is more than one sequence

Reply:
Yeah, that makes sense. I will double check that and do either 1st or 3rd from your suggestions.

Reply:
I think we discussed last week in the design meeting the fact that there was many help desk requests about submitting multiple blast jobs at once, which would require validation of multiple sequences in one go and then sequential submissions. We are not implementing sequential submissions in the first attempt, but it will likely be required in the future.
So maybe suggestions #2 is best

Reply:
So in order to keep the function input and return consistent, should I make this function to always accept and array of sequences and always return an array of results too? Only sometimes returning an array can force the user of function to write more code.

Reply:
That would be sensible. For me OK to return always arrays

Reply:
I pushed another commit to handle array of sequences, but it still needs to handle a FASTA string and return an array of results, so please wait until that one is pushed too as well.